### PR TITLE
fix for different issues

### PR DESCRIPTION
**Pull request by stargieg**
Client timeouts
Wrong reply
Duplicate uci option after opkg update
No listen if netif is hotplug
Overwritten conffile after opkg update

Signed-off-by: Patrick Grimm <patrick@lunatiki.de>

### CODE REVIEW

**Comment by stargieg** [comment]
FYI: here is my setup kadnode wizard script. works so far well. :-)
https://github.com/stargieg/luci2-ffwizard/blob/master/luci2-ffwizard/files/etc/ffwizard.d/60-kadnode.sh

**Comment by mwarning** [comment]
Ah wow, thanks a lot. I will review it as soon as i can.

File: openwrt/kadnode/files/kadnode.init [hunk 1]
```diff
@@ -85,6 +93,23 @@ stop_instance() {
 	rm -f $CONFIG_FILE
 }
 
+add_interface_trigger()
+{
```
Comment (mwarning) [inline openwrt/kadnode/files/kadnode.init]: Put the `{` behind the function name. This keeps the code style consistent.

File: openwrt/kadnode/files/kadnode.init [hunk 2]
```diff
@@ -85,6 +93,23 @@ stop_instance() {
...
+	local interface
+
+	config_get ifname "$1" ifname
```
Comment (mwarning) [inline openwrt/kadnode/files/kadnode.init]: `interface` does not seem to get set.

File: src/ext-dns.c [hunk 3]
```diff
@@ -499,6 +494,9 @@ static int dns_setup_msg(struct Message *msg, const struct result_t *results, co
 	msg->nsCount = 0;
 	msg->arCount = 0;
 
+	if (!results) {
+		return 1;
```
Comment (mwarning) [inline src/ext-dns.c]: With this check, we do not need to later checks in the for statement anymore. We can remove those.

File: src/ext-dns.c [hunk 4]
```diff
@@ -512,20 +510,22 @@ static int dns_setup_msg(struct Message *msg, const struct result_t *results, co
 
 		for (result = results, i = 0; result && c < MAX_ADDR_RECORDS; result = result->next) {
 			if (is_valid_result(result)) {
-				setAddressRecord(&msg->answers[c++], g_names[i++], &result->addr);
```
Comment (mwarning) [inline src/ext-dns.c]: hm, I would need to check if g_names was intentional here. (g_names are placeholder names needed in service records iirc)

File: openwrt/kadnode/files/kadnode.init (hunk 1 above)
Comment (stargieg) [reply openwrt/kadnode/files/kadnode.init]: Ack see 4a61225262f82b8b52f7fc05da0e01e61d0dab8e

File: openwrt/kadnode/files/kadnode.init (hunk 2 above)
Comment (stargieg) [reply openwrt/kadnode/files/kadnode.init]: Ack see 4a61225262f82b8b52f7fc05da0e01e61d0dab8e

File: src/ext-dns.c (hunk 3 above)
Comment (stargieg) [reply src/ext-dns.c]: I don't know what you mean.
If "results" is empty the "for" loops do not have to be processed. This is for speedup.

File: src/ext-dns.c (hunk 4 above)
Comment (stargieg) [reply src/ext-dns.c]: Yes, g_names is only needed for SRV setServiceRecord() records rr->rd_data.srv_record.target

File: openwrt/kadnode/files/kadnode.init [hunk 5]
```diff
@@ -85,6 +93,22 @@ stop_instance() {
...
+add_interface_trigger() {
+	local ifname
+
+	config_get ifname "$1" ifname
+
+	[ -n "$ifname" ] && procd_add_interface_trigger "interface.*" "$ifname" /etc/init.d/kadnode restart
+}
+
+service_triggers()
+{
```
Comment (mwarning) [inline openwrt/kadnode/files/kadnode.init]: here is also a stray {

File: src/ext-dns.c (hunk 4 above)
Comment (mwarning) [reply src/ext-dns.c]: it looks like this code section is dealing with service records..

File: src/ext-dns.c [hunk 6]
```diff
@@ -546,6 +546,18 @@ static const char* qtype_str(int qType)
...
+	case CNAME_Resource_RecordType:
+		return "CNAME";
+	case SOA_Resource_RecordType:
+		return "SOA";
+	case MX_Resource_RecordType:
+		return "MX";
+	case TXT_Resource_RecordType:
+		return "TXT";
+	case ANY_Resource_RecordType:
+		return "ANY";
```
Comment (mwarning) [inline src/ext-dns.c]: hm, these entry type are mostly not supported anyway. I would like to keep the code minimal.

File: src/ext-dns.c [hunk 7]
```diff
@@ -676,18 +687,6 @@ static void dns_handler(int rc, int sock)
 		return;
 	}
 
-	if (af != AF_UNSPEC) {
```
Comment (mwarning) [inline src/ext-dns.c]: Was there a problem with this code? It was making sure that we do not handle request that we cannot answer.

File: src/ext-dns.c [hunk 8]
```diff
@@ -697,30 +696,39 @@ static void dns_handler(int rc, int sock)
...
+			if (dns_setup_msg(&msg, search->results, NULL) < 1) {
+				log_debug("DNS: Failed to setup msg");
+				dns_setup_msg(&msg, NULL, NULL);
+			}
 		}
 
 		log_debug("DNS: Send back addresses to: %s", str_addr(&clientaddr));
+	} else {
+		log_warning("DNS: No question for A, AAAA, SRV, PTR or ANY resource found in query.");
+		dns_setup_msg(&msg, NULL, NULL);
```
Comment (mwarning) [inline src/ext-dns.c]: It looks like you want to make sure that there is send an answer (even if there is no response) - that is good.

File: src/ext-dns.c [hunk 9]
```diff
@@ -697,30 +696,39 @@ static void dns_handler(int rc, int sock)
...
+			}
 		}
 
 		log_debug("DNS: Send back hostname '%s' to: %s",
 			domain, str_addr(&clientaddr)
 		);
-	} else {
+	} else if (msg.question.qType == A_Resource_RecordType
+		|| msg.question.qType == AAAA_Resource_RecordType
+		|| msg.question.qType == ANY_Resource_RecordType
```
Comment (mwarning) [inline src/ext-dns.c]: What about SRV requests?

Comment (mwarning) [reply src/ext-dns.c]: And using ANY might not be a good direction: https://blog.cloudflare.com/rfc8482-saying-goodbye-to-any/

File: openwrt/kadnode/files/kadnode.init (hunk 5 above)
Comment (stargieg) [reply openwrt/kadnode/files/kadnode.init]: ACK see 533f8210e39a5e5e48607a0ef80aafaa82bbdf18

File: src/ext-dns.c (hunk 6 above)
Comment (stargieg) [reply src/ext-dns.c]: ACK see c59d397c21100be53b3ca1a2bb222c9ee6d138b6
and the background https://blog.cloudflare.com/rfc8482-saying-goodbye-to-any/

**Comment by mwarning** [comment]
welcome back. I will try to finish the review next week. Anyway, it seems the main issues seems to be a missing reply even when there was now result found. That's good to be fixed.

File: src/ext-dns.c [hunk 10]
```diff
@@ -512,20 +509,22 @@ static int dns_setup_msg(struct Message *msg, const struct result_t *results, co
...
+	} else if (msg->question.qType == PTR_Resource_RecordType && hostname != NULL) {
 		setPointerRecord(&msg->answers[c++], qName, hostname);
 		msg->anCount++;
 		i = 1;
 	} else {
 		// Assume AAAA or A Record Type
 		for (result = results, i = 0; result && c <= MAX_ADDR_RECORDS; result = result->next) {
 			if (is_valid_result(result)) {
-				setAddressRecord(&msg->answers[c++], g_names[i++], &result->addr);
+				setAddressRecord(&msg->answers[c++], qName, &result->addr);
```
Comment (mwarning) [inline src/ext-dns.c]: this change is still puzzling me a bit. Need to think about it tomorrow.

File: src/ext-dns.c (hunk 4 above)
Comment (mwarning) [reply src/ext-dns.c]: g_names[i++] are placeholder names. It's a reference into the service records added in the previous loop.

The service record contains the actual host name and port.
I do this to included the port in the answer.

Can remove that change?

File: src/ext-dns.c (hunk 7 above)
Comment (mwarning) [reply src/ext-dns.c]: I see, since you want to always return a answer, you removed this early return.

File: src/ext-dns.c (hunk 3 above)
Comment (mwarning) [reply src/ext-dns.c]: I think this change is ok. I felt that having only one exit in a method when possible is good, as it is less likely to skip initialization of data by accident.

**Comment by mwarning** [comment]
Ok, seems that there is only one line change that needs to be reverted for this PR to got through.

**Comment by mwarning** [comment]
Merging now before the changes get stale. It is always a good idea to submit changes in multiple PRs in the future, so that one PR does not block the others.

**Pull request merged**